File: message-java/src/test/java/quan/message/role/SRoleLogin.java
package quan.message.role;

import quan.message.Buffer;
import java.util.*;
import quan.message.Message;
import quan.message.user.UserInfo;
import java.io.IOException;

/**
 * 角色登录
 * Created by 自动生成
 */
public class SRoleLogin extends Message {

    private long roleId = 0L;//角色id

    private String roleName = "";//角色名

    private RoleInfo roleInfo;//角色信息

    private ArrayList<RoleInfo> roleInfoList = new ArrayList<>();//角色信息

    private HashSet<RoleInfo> roleInfoSet = new HashSet<>();//角色信息

    private HashMap<Long, RoleInfo> roleInfoMap = new HashMap<>();//角色信息

    private UserInfo userInfo = new UserInfo();//用户信息

    public SRoleLogin() {
        super(2222);
    }

    public long getRoleId() {
        return roleId;
    }

    public SRoleLogin setRoleId(long roleId) {
        this.roleId = roleId;
        return this;
    }

    public String getRoleName() {
        return roleName;
    }

    public SRoleLogin setRoleName(String roleName) {
        if (roleName == null){
            throw new NullPointerException();
        }
        this.roleName = roleName;
        return this;
    }

    public RoleInfo getRoleInfo() {
        return roleInfo;
    }

    public SRoleLogin setRoleInfo(RoleInfo roleInfo) {
        this.roleInfo = roleInfo;
        return this;
    }

    public ArrayList<RoleInfo> getRoleInfoList() {
        return roleInfoList;
    }

    public HashSet<RoleInfo> getRoleInfoSet() {
        return roleInfoSet;
    }

    public HashMap<Long, RoleInfo> getRoleInfoMap() {
        return roleInfoMap;
    }

    public UserInfo getUserInfo() {
        return userInfo;
    }

    public SRoleLogin setUserInfo(UserInfo userInfo) {
        if (userInfo == null){
            throw new NullPointerException();
        }
        this.userInfo = userInfo;
        return this;
    }

    @Override
    public SRoleLogin create() {
        return new SRoleLogin();
    }

    @Override
    public void encode(Buffer buffer) throws IOException {
        super.encode(buffer);

        buffer.writeLong(roleId);
        buffer.writeString(roleName);

        buffer.writeBool(roleInfo != null);
        if (roleInfo != null) {
            roleInfo.encode(buffer);
        }

        buffer.writeInt(roleInfoList.size());
        for (RoleInfo roleInfoListValue : roleInfoList) {
            roleInfoListValue.encode(buffer);
        }

        buffer.writeInt(roleInfoSet.size());
        for (RoleInfo roleInfoSetValue : roleInfoSet) {
            roleInfoSetValue.encode(buffer);
        }

        buffer.writeInt(roleInfoMap.size());
        for (long roleInfoMapKey : roleInfoMap.keySet()) {
            buffer.writeLong(roleInfoMapKey);
            roleInfoMap.get(roleInfoMapKey).encode(buffer);
        }

        userInfo.encode(buffer);
    }

    @Override
    public void decode(Buffer buffer) throws IOException {
        super.decode(buffer);

        roleId = buffer.readLong();
        roleName = buffer.readString();

        if (buffer.readBool()) {
            if (roleInfo == null) {
                roleInfo = new RoleInfo();
            }
            roleInfo.decode(buffer);
        }

        int _roleInfoList_Size = buffer.readInt();
        for (int i = 0; i < _roleInfoList_Size; i++) {
            RoleInfo _roleInfoList_Value = new RoleInfo();
            _roleInfoList_Value.decode(buffer);
            roleInfoList.add(_roleInfoList_Value);
        }

        int _roleInfoSet_Size = buffer.readInt();
        for (int i = 0; i < _roleInfoSet_Size; i++) {
            RoleInfo _roleInfoSet_Value = new RoleInfo();
            _roleInfoSet_Value.decode(buffer);
            roleInfoSet.add(_roleInfoSet_Value);
        }

        int _roleInfoMap_Size = buffer.readInt();
        for (int i = 0; i < _roleInfoMap_Size; i++) {
            long _roleInfoMap_Key = buffer.readLong();
            RoleInfo _roleInfoMap_Value = new RoleInfo();
            _roleInfoMap_Value.decode(buffer);
            roleInfoMap.put(_roleInfoMap_Key, _roleInfoMap_Value);
        }

        userInfo.decode(buffer);
    }

    @Override
    public String toString() {
        return "SRoleLogin{" +
                "roleId=" + roleId +
                ",roleName='" + roleName + '\'' +
                ",roleInfo=" + roleInfo +
                ",roleInfoList=" + roleInfoList +
                ",roleInfoSet=" + roleInfoSet +
                ",roleInfoMap=" + roleInfoMap +
                ",userInfo=" + userInfo +
                '}';

    }

}

File: message/message-cs/src/quan/message/role/SRoleLogin.cs
using System;
using System.Collections.Generic;
using quan.message;
using quan.message.user;

namespace quan.message.role
{
    public class SRoleLogin : Message
    {
		public long roleId{ get; set; }

		private string _roleName = "";

		public string roleName
		{
	    	get => _roleName;
	    	set => _roleName = value ?? throw new NullReferenceException();
		}

		private RoleInfo _roleInfo = new RoleInfo();

		public RoleInfo roleInfo
		{
	    	get => _roleInfo;
	    	set => _roleInfo = value ?? throw new NullReferenceException();
		}

		public List<RoleInfo> roleInfoList { get; } = new List<RoleInfo>();

		public HashSet<RoleInfo> roleInfoSet { get; } = new HashSet<RoleInfo>();

		public Dictionary<long, RoleInfo> roleInfoMap { get; } = new Dictionary<long, RoleInfo>();

		private UserInfo userInfo;


		public SRoleLogin(): base(222)
		{
		}

		public override Message create()
		{
			return new SRoleLogin();
		}


		public override void Encode(Buffer buffer)
		{
	    	base.Encode(buffer);

		    buffer.WriteLong(roleId);
		    buffer.WriteString(roleName);
		    roleInfo.Encode(buffer);

		    buffer.WriteInt(roleInfoList.Count);
		    foreach (var _roleInfoList_Value in roleInfoList) {
			    _roleInfoList_Value.Encode(buffer);
		    }

		    buffer.WriteInt(roleInfoSet.Count);
		    foreach (var _roleInfoSet_Value in roleInfoSet) {
			    _roleInfoSet_Value.Encode(buffer);
		    }

		    buffer.WriteInt(roleInfoMap.Count);
		    foreach (var _roleInfoMap_Key in roleInfoMap.Keys) {
		        buffer.WriteLong(_roleInfoMap_Key);
			    roleInfoMap[_roleInfoMap_Key].Encode(buffer);
		    }

		    buffer.WriteBool(userInfo != null);
		    if (userInfo != null) {
		        userInfo.Encode(buffer);
		    }

		}

		public override void Decode(Buffer buffer)
		{
	    	base.Decode(buffer);

		    roleId = buffer.ReadLong();
		    roleName = buffer.ReadString();
		    roleInfo.Decode(buffer);

		    int _roleInfoList_Size = buffer.ReadInt();
		    for (int i = 0; i < _roleInfoList_Size; i++) {
			    RoleInfo _roleInfoList_Value = new RoleInfo();
			    _roleInfoList_Value.Decode(buffer);
			    roleInfoList.Add(_roleInfoList_Value);
		    }

		    int _roleInfoSet_Size = buffer.ReadInt();
		    for (int i = 0; i < _roleInfoSet_Size; i++) {
			    RoleInfo _roleInfoSet_Value = new RoleInfo();
			    _roleInfoSet_Value.Decode(buffer);
			    roleInfoSet.Add(_roleInfoSet_Value);
		    }

		    int _roleInfoMap_Size = buffer.ReadInt();
		    for (int i = 0; i < _roleInfoMap_Size; i++) {
			    long _roleInfoMap_Key = buffer.ReadLong();
			    RoleInfo _roleInfoMap_Value = new RoleInfo();
			    _roleInfoMap_Value.Decode(buffer);
			    roleInfoMap.Add(_roleInfoMap_Key, _roleInfoMap_Value);
		    }

		    if (buffer.ReadBool()) {
		        if (userInfo == null) {
		            userInfo = new UserInfo();
		        }
		        userInfo.Decode(buffer);
            }

		}
    }
}